Run existing pool DB property tests on model implementation.

jonathanknowles opened the pull request:
# Issue Number

https://jira.iohk.io/browse/ADP-406

# Overview

Over time, we've built up a fairly large suite of [property tests for the pool database](https://github.com/input-output-hk/cardano-wallet/blob/ff181966317b50083a61aa3e4f3dfd2ac1bff1a3/lib/core/test/unit/Cardano/Pool/DB/Properties.hs).

However, up until now these tests were only ever run on the **SQLite** implementation. The **model** implementation was not tested.

This PR fixes this by adjusting the test suite to test both the **SQLite** and **model** implementations. For example:

```hs
ghcid \
    --command "stack ghci cardano-wallet-core:lib cardano-wallet-core:test:unit" \
    --test ':run Spec.main -m prop_listPoolLifeCycleData_multiplePools_multipleCerts'

Cardano.Pool.DB.MVar
  MVar
    Stake Pool Properties
      prop_listPoolLifeCycleData_multiplePools_multipleCerts
        +++ OK, passed 100 tests.
Cardano.Pool.DB.Sqlite
  Sqlite
    Stake Pool Properties
      prop_listPoolLifeCycleData_multiplePools_multipleCerts
        +++ OK, passed 100 tests.

Finished in 40.3082 seconds
2 examples, 0 failures
```

## Details

This PR:
1. Generalizes the `Pool.DB.Properties` module to work with any implementation of `Pool.DBLayer`.
2. Adds a `Pool.DB.MVarSpec` module to handle the required setup for testing the model implementation.
3. Fixes the small number of resulting failures.

jonathanknowles committed:
Move SQLite-specific testing functions into `SqliteSpec`.

The `Properties` module now only contains functions that are general to
all implementations of `Pool.DB.DBLayer`.

jonathanknowles committed:
Test the model implementation of the pool database.

`Cardano.Pool.DB.Properties` already provides a large set of properties
that can be run against any implementation of `Cardano.Pool.DB.DBLayer`.

This commit arranges for these properties to be tested against the model
DB implementation provided by `Cardano.Pool.DB.MVar`.

(Prior to this commit, we only tested the SQLite implementation.)

jonathanknowles committed:
Fix the model implementation of `listRetiredPools`.

The `prop_poolRetirement` property expects that `listRetiredPools` will
include all pools with non-cancelled retirement certificates, even when
those pools never had registration certificates to begin with.

Strictly speaking, this expectation is more than we need, as in reality
we'll probably never see a case where a pool publishes a retirement
certificate without there being a prior registration certificate. (This
should be explicitly forbidden by the ledger layer.)

However, it turns out to be more convenient to express the property this
way, as:

1. It makes the property much simpler to express. We don't need to worry
   about registration certificates in the setup phase.
2. It makes the SQLite query for implementing `listRetiredPools` much
   easier to write.

Therefore, in this case, we simply update the model to match the
behaviour of our SQLite implementation.

jonathanknowles committed:
Remove unnecessary ordering restriction from `prop_listRegisteredPools`.

The `listRegisteredPools` operation doesn't guarantee to list pools in
any particular order, so our property tests also shouldn't require any
particular order.

jonathanknowles committed:
Fix the model implementation of `mUnfetchedPoolMetadataRefs`.

The property expects that recent fetch attempts are excluded from the
list returned by `unfetchedPoolMetadataRefs`. We adjust the model to
satisfy the property.

Note that the property coverage here isn't necessarily as complete as it
could be. We'd ideally want to test this pool DB function in conjunction
with other pool DB functions as part of a quickcheck-state-machine style
test.

jonathanknowles committed:
Use function `retainOnlyMostRecent` to clarify `mListRetiredPools`.

It's not entirely obvious that `Map.mapKeys` retains only the value
corresponding to the highest original key.

We add a comment to make this clear.

rvl commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -935,7 +911,7 @@ prop_listRegisteredPools DBLayer {..} entries =
             [ "Read from DB: " <> show pools
             ]
-        assert (pools == (view #poolId <$> reverse entries))
+        assert (Set.fromList pools == Set.fromList (view #poolId <$> entries))
```
Any reason to remove the check on ordering?

jonathanknowles replied:
> Any reason to remove the check on ordering?

Good question.

The function `listRegisteredPools` doesn't guarantee (currently) to list pools in any particular order, so it doesn't make sense for the property to expect a particular order.

(And as it turns out, the ordering of pools returned by the SQLite implementation of `listRegisteredPools` really is different from the ordering we get back from `Set.toList`. It seems unnecessary to make these orders the same by adding an extra ascending sort to the SQLite implementation.)

rvl reviewed (approved):
It can't hurt to test the tests more.

jonathanknowles committed:
Use `List.sort` instead of `Set.fromList`.

The `listRegisteredPools` function doesn't guarantee any particular
order, but it should guarantee to not return any duplicate elements.

Use `List.sort` here instead of `Set.fromList` to avoid masking this
potential error.

(Alternatively, we could just adjust the type of `listRegisteredPools`
to return a set instead of a list, but that is a more invasive change.)

jonathanknowles committed:
Remove unnecessary intermediate assignment.

jonathanknowles committed:
Use `assertWith` to clarify property `prop_listRegisteredPools`.
